src/map.rs

```diff
@@ -532,6 +532,205 @@ where
+                return None;
+            }
+
+            // slow path -- bin is non-empty
+            // safety: bin is a valid pointer.
+            //
+            // there are two cases when a bin pointer is invalidated:
+            //
+            //  1. if the table was resized, bin is a move entry, and the resize has completed. in
+            //     that case, the table (and all its heads) will be dropped in the next epoch
+            //     following that.
+            //  2. if the table is being resized, bin may be swapped with a move entry. the old bin
+            //     will then be dropped in the following epoch after that happens.
+            //
+            // in both cases, we held the guard when we got the reference to the bin. if any such
+            // swap happened, it must have happened _after_ we read. since we did the read while
+            // pinning the epoch, the drop must happen in the _next_ epoch (i.e., the one that we
+            // are holding up by holding on to our guard).
+            match *unsafe { bin.deref() } {
+                BinEntry::Moved(next_table) => {
+                    table = self.help_transfer(table, next_table, guard);
+                }
+                BinEntry::Node(ref head) => {
+                    // bin is non-empty, need to link into it, so we must take the lock
+                    let head_lock = head.lock.lock();
+
+                    // need to check that this is _still_ the head
+                    let current_head = t.bin(bini, guard);
+                    if current_head != bin {
+                        // nope -- try again from the start
+                        continue;
+                    }
+
+                    // yes, it is still the head, so we can now "own" the bin
+                    // note that there can still be readers in the bin!
+
+                    // TODO: TreeBin & ReservationNode
+
+                    let mut removed_node = false;
+                    let mut bin_count = 1;
+                    let mut p = bin;
+                    let mut pred: Shared<'_, BinEntry<K, V>> = Shared::null();
+
+                    let new_val = loop {
+                        // safety: we read the bin while pinning the epoch. a bin will never be
+                        // dropped until the next epoch after it is removed. since it wasn't
+                        // removed, and the epoch was pinned, that cannot be until after we drop
+                        // our guard.
+                        let n = unsafe { p.deref() }.as_node().unwrap();
+                        // TODO: This Ordering can probably be relaxed due to the Mutex
+                        let next = n.next.load(Ordering::SeqCst, guard);
+                        if n.hash == h && n.key.borrow() == key {
+                            // the key already exists in the map!
+                            let current_value = head.value.load(Ordering::SeqCst, guard);
```

Comment:
Why is `current_value` is loaded from `head.value` instead of `n.value`?

Reply:
That is an excellent question! I believe that is a bug. I'll push a fix and a unit test for that shortly!

Reply:
Fixed in https://github.com/jonhoo/flurry/commit/051ca79fe2458af0e0e75ab93f8cd113078cd2bd